app/gui/src/dashboard/providers/LocalStorageProvider.tsx

```diff
@@ -41,61 +40,106 @@ export default function LocalStorageProvider(props: LocalStorageProviderProps) {
+    LocalStorage.getInstance().delete(key)
+  }
+
+  const get = (localStorage: LocalStorage = LocalStorage.getInstance()) => {
+    // eslint-disable-next-line no-restricted-syntax, @typescript-eslint/no-unsafe-return
+    return localStorage.get(key) as Value | undefined
+  }
+
+  const set = (value: Value, localStorage: LocalStorage = LocalStorage.getInstance()) => {
+    localStorage.set(key, value)
+  }
+
+  const deleteKey = (localStorage: LocalStorage = LocalStorage.getInstance()) => {
+    localStorage.delete(key)
+  }
+
+  const useGet = () => {
+    const { localStorage } = useLocalStorage()
+
+    useEffect(() => {
+      if ('error' in schema.safeParse(localStorage.get(key))) {
+        localStorage.delete(key)
+      }
+    }, [localStorage])
+
+    // The return type is not `any`, this is a bug in ESLint.
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+    return useEventCallback(() => get(localStorage))
+  }
+
+  const useSet = () => {
+    const { localStorage } = useLocalStorage()
+    return useEventCallback((value: Value) => {
+      set(value, localStorage)
+    })
+  }
+
+  const useDelete = () => {
+    const { localStorage } = useLocalStorage()
+    return useEventCallback(() => {
+      deleteKey(localStorage)
+    })
+  }
+
+  function useLocalStorageState(): readonly [
+    value: Value | undefined,
+    setValue: (newValue: SetStateAction<Value | undefined>) => void,
+  ]
+  function useLocalStorageState(
+    defaultValue: Value,
+  ): readonly [value: Value, setValue: (newValue: SetStateAction<Value>) => void]
+  /** Subscribe to Local Storage updates for a specific key. */
+  // eslint-disable-next-line no-restricted-syntax
+  function useLocalStorageState(
+    defaultValue?: Value,
+  ): readonly [
+    value: Value | undefined,
+    setValue: (newValue: SetStateAction<Value | undefined>) => void,
+  ] {
+    const { localStorage } = useLocalStorage()
+
+    const [value, privateSetValue] = useState<Value | undefined>(
+      () => get(localStorage) ?? defaultValue,
+    )
+
+    const setValue = useEventCallback((newValue: SetStateAction<Value | undefined>) => {
       privateSetValue((currentValue) => {
-        const nextValue = typeof newValue === 'function' ? newValue(currentValue) : newValue
-
+        const nextValue = isFunction(newValue) ? newValue(currentValue) : newValue
         if (nextValue === undefined) {
           localStorage.delete(key)
         } else {
           localStorage.set(key, nextValue)
         }
-
         return nextValue
       })
-    },
-  )
-
-  React.useEffect(
-    () =>
-      localStorage.subscribe(key, (newValue) => {
-        privateSetValue(newValue ?? defaultValue)
-      }),
-    [defaultValue, key, localStorage],
-  )
+    })
 
-  return [value, setValue]
+    return [value, setValue]
+  }
+  return { get, set, delete: deleteKey, useGet, useSet, useDelete, useState: useLocalStorageState }
```

Comment:
can we unify the stuff here, like:

```typescript
// use array to make it easier to rename
return [
  item: {
    get,set,clear
  },
  useItem: useLocalStorageState(): [value, setValue, clearValue]
]
```

Hook might be less performant but more convinient to use.
For perf reasons we might expose setter as third item:

```typescript
...
useSetItem(): [setValue, clearValue]
```

Reply:
> `use array to make it easier to rename`

not sure about this one, especially since you don't always need both non-reactive and reactive parts.

i think unifying the values as you suggested should be enough

Reply:
> Same as for useState

i know `useState` does it but IMO that's because a getter/setter pair is a very common pattern for reactive state - and because you almost always want to rename the accessors anyway

Reply:
And we dont want to rename the result of 'defineLocalStorageItem'?

Reply:
you do have a point! still not 100% sure as we don't always want all the members and it's a lot less obvious what each returned value is (it's not a popular pattern unlike get/set), but if you think it's better than the current option then we can absolutely do that